alf/environments/carla_sensors.py

```diff
@@ -1393,3 +1413,399 @@ def get_next_waypoint_index(self):
+                vehicle (ev) to the bottom of the BEV image.
+            history_idx (list[int]): a list of numbers representing the indices
+                of the history information to be rendered for non-ego vehicles
+            max_history_len (int): max number of history length preserved
+            vehicle_bbox_factor (float): a factor to scale the vehicle bounding
+                boxes
+            walker_bbox_factor (float): a factor to scale the walker bounding
+                boxes
+        """
+
+        super().__init__(parent_actor)
+        self._width = image_size_in_pixels
+        self._pixels_ev_to_bottom = pixels_ev_to_bottom
+        self._pixels_per_meter = pixels_per_meter
+        self._history_idx = history_idx
+        self._vehicle_bbox_factor = vehicle_bbox_factor
+        self._walker_bbox_factor = walker_bbox_factor
+        self._use_rgb_image = use_rgb_image
+        self._history_queue = deque(maxlen=max_history_len)
+
+        self._image_channels = 3
+        self._masks_channels = 3 + 3 * len(self._history_idx)
+
+        self._alf_world = alf_world
+        self._world = alf_world._world
+        self._navigation_sensor = navigation_sensor
+        self._map_handler = MapHandler(self._world, pixels_per_meter)
+        map_masks = self._map_handler.get_masks()
+        self._road = map_masks["road"]
+        self._lane = map_masks["lane"]
+
+        # register traffic lights
+        TrafficLightHandler.reset(self._alf_world)
+
+        self._distance_threshold = np.ceil(
+            self._width / self._pixels_per_meter)
+
+        if self._use_rgb_image:
+            self._observation_spec = alf.TensorSpec(
+                [3, self._width, self._width], dtype='uint8')
+        else:
+            self._observation_spec = alf.TensorSpec(
+                [self._masks_channels, self._width, self._width],
+                dtype='uint8')
+
+    def observation_spec(self):
+        return self._observation_spec
+
+    def observation_desc(self):
+        return ("3-D vector of [bev_channels, height, width].")
+
+    def _get_actor_bounding_box(self, actor_filter):
+        bounding_boxes = []
+        for vehicle in self._world.get_actors().filter(actor_filter):
+            bounding_boxes.append(
+                (vehicle.get_location(), vehicle.bounding_box,
+                 vehicle.get_transform()))
+        return bounding_boxes
+
+    def _generate_observation(self, render_rgb_image):
+
+        ev_transform = self._parent.get_transform()
+        ev_loc = ev_transform.location
+        ev_rot = ev_transform.rotation
+        ev_bbox = self._parent.bounding_box
+
+        def is_within_distance(loc):
+            c_distance = abs(ev_loc.x - loc.x) < self._distance_threshold \
+                and abs(ev_loc.y - loc.y) < self._distance_threshold \
+                and abs(ev_loc.z - loc.z) < 8.0
+            c_ev = abs(ev_loc.x - loc.x) < 1.0 and abs(ev_loc.y - loc.y) < 1.0
```

Comment:
Is this to rule out the ego vehicle itself? If would be good to document this. Also it would be slightly more performant to cache the x/y difference, and rule out `c_ev` case before computing `c_distance`.

Reply:
Correct, this is to determine whether it is ego-vehicle.  Added documentation, and changed to rule out ``c_ev`` case first.